django/db/migrations/graph.py

```diff
@@ -108,27 +131,101 @@ def __init__(self):
-                child
+            error_message = (
+                "Migration %s dependencies reference nonexistent"
+                " child node %r" % (migration, child)
             )
+            self.add_dummy_node(child, migration, error_message)
         if parent not in self.nodes:
-            raise NodeNotFoundError(
-                "Migration %s dependencies reference nonexistent parent node %r" % (migration, parent),
-                parent
+            error_message = (
+                "Migration %s dependencies reference nonexistent"
+                " parent node %r" % (migration, parent)
             )
+            self.add_dummy_node(parent, migration, error_message)
         self.node_map[child].add_parent(self.node_map[parent])
         self.node_map[parent].add_child(self.node_map[child])
+        if not skip_validation:
+            self.validate_consistency()
         self.clear_cache()
 
+    def replace_node(self, key, replacement):
+        """
+        If `key` exists, it is removed and any dependencies that were
+        referencing it are changed to reference the `replacement` node
+        instead.
+        """
+        self.nodes.pop(key, None)
+        node = self.node_map.pop(key, None)
+        if node:
+            replacement_node = self.node_map[replacement]
+            for child in node.children:
+                child.parents.remove(node)
+                if child.key != replacement:
+                    replacement_node.add_child(child)
+                    child.add_parent(replacement_node)
+            for parent in node.parents:
+                parent.children.remove(node)
+                if parent.key != replacement:
+                    replacement_node.add_parent(parent)
+                    parent.add_child(replacement_node)
+
+    def remove_replacement_node(self, key, replaced_nodes):
+        """
+        Think of this as the inverse operation to `replace_node`. Kind of.
+        Removes the replacement node `key` and remaps its child nodes to
+        `replaced_nodes` - the list of nodes it would have replaced.
+        """
+        self.nodes.pop(key, None)
+        node = self.node_map.pop(key, None)
+        for child in node.children:
+            child.parents.remove(node)
+            for replaced in replaced_nodes:
+                # TODO: This needs a performance optimisation.
+                # If the replaced nodes all live in the same app, then only the
+                # latest one needs to have the replacing migration's children
+                # attached to it. Currently ALL of the replaced nodes will
+                # get these children. Not incorrect, just slow.
+                replaced_node = self.node_map.get(replaced)
+                if replaced_node:
+                    replaced_node.add_child(child)
+                    child.add_parent(replaced_node)
+        for parent in node.parents:
+            parent.children.remove(node)
```

Comment:
Oh, yeah. That makes sense. Can you leave a comment with an explanation instead of adding that code, please.